### PR TITLE
SecureBackup: update matrix sdk module.

### PR DESCRIPTION
**Pull request by bmarty**
SDK part of #1648.

### CODE REVIEW

File: libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.libraries.matrix.api.encryption
+
+import kotlinx.coroutines.flow.StateFlow
+
+interface EncryptionService {
+    val backupStateStateFlow: StateFlow<BackupState>
```
Comment (julioromano) [inline libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt:22]: Just style:
`StateState` :/
unless we have a very strict naming scheme we want to stick to I'd suggest we could drop the `StateFlow` suffix and just use `Flow` even though it doesn't match the actual type name.

File: libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    val recoveryStateStateFlow: StateFlow<RecoveryState>
+    val backupUploadStateStateFlow: StateFlow<BackupUploadState>
+    val enableRecoveryProgressStateFlow: StateFlow<EnableRecoveryProgress>
+
+    suspend fun enableBackups(): Result<Unit>
```
Comment (julioromano) [inline libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt:27]: Is there any way to disable them?

File: libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    suspend fun isLastDevice(): Result<Boolean>
```
Comment (julioromano) [inline libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt:29]: Missing context: last what? Maybe add docstring

File: libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            service.enableRecovery(
+                waitForBackupsToUpload = waitForBackupsToUpload,
+                progressListener = object : EnableRecoveryProgressListener {
+                    override fun onUpdate(status: RustEnableRecoveryProgress) {
+                        enableRecoveryProgressStateFlow.value = enableRecoveryProgressMapper.map(status)
+                    }
+                }
+            )
+                // enableRecovery returns the encryption key, but we read it from the state flow
+                .let { }
```
Comment (julioromano) [inline libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt:93]: Why `.let{}` when `Unit` was enough ?

File: libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    fun destroy() {
+        // No way to remove the listeners...
+        service.destroy()
+    }
+
+    override suspend fun enableBackups(): Result<Unit> = withContext(dispatchers.io) {
+        runCatching {
+            service.enableBackups()
```
Comment (julioromano) [inline libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt:76]: Just a question, why none of these ffi methods are `suspend`? Didn't we just get rid of all `*blocking` versions of many APIs?

**Review by julioromano — APPROVED** [review]
LGTM, just nits.

File: libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt (hunk 1 above)
Comment (bmarty) [reply libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt:22]: Yes, I agree it's not ideal... I will change it in a coming PR.

File: libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt (hunk 4 above)
Comment (bmarty) [reply libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt:93]: `Unit` is working, but quality check does not like it.

Comment (julioromano) [reply libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt:93]: Why doesn't it? It's the most correct thing to do in this case kotlin-wise.

File: libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt (hunk 2 above)
Comment (bmarty) [reply libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt:27]: Yes, there is one, but the name from the SDK is `disableRecovery`, do not ask me why. I will document the API later.
We have to decide if we stick to the SDK name or if we change it to make is clearer.

File: libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt (hunk 3 above)
Comment (bmarty) [reply libraries/matrix/api/src/main/kotlin/io/element/android/libraries/matrix/api/encryption/EncryptionService.kt:29]: This is the last Session, so last device that logged in. Will document the API.

File: libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt (hunk 4 above)
Comment (bmarty) [reply libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt:93]: Yeah, I agree. Detekt is reporting this:
> ~/workspaces/element-x-android/libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt:84:13: A single Unit expression is unnecessary and can safely be removed. [OptionalUnit]

There are several similar reports in the detekt project https://github.com/detekt/detekt, but all the issues seems to be close.

In the mean time, using `.let { }` is the solution, and it's already used in the codebase.

Comment (julioromano) [reply libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt:93]: Seems like a detekt bug

**Pull request merged**